core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala

```diff
@@ -80,94 +80,49 @@ case object FilterDustFinalizer extends RawTxFinalizer {
-    }
-
-    val dummyTxF = TxUtil.addDummySigs(txNoChangeFee, inputInfos)
+    val tx = txBuilderResult.toBaseTransaction
 
-    val txF = dummyTxF.map { dummyTx =>
-      val fee = feeRate.calc(dummyTx)
-      val change = totalCrediting - totalSpending - fee
-      val newChangeOutput = TransactionOutput(change, changeSPK)
-      val newOutputs = if (change <= Policy.dustThreshold) {
-        outputs
-      } else {
-        outputs :+ newChangeOutput
-      }
+    val passInOutChecksT =
+      SanityCheckFinalizer.sanityDestinationChecks(inputInfos.map(_.outPoint),
+                                                   expectedOutputs,
+                                                   changeSPKs,
+                                                   tx)
 
-      txNoChangeFee match {
-        case btx: NonWitnessTransaction =>
-          BaseTransaction(btx.version, btx.inputs, newOutputs, btx.lockTime)
-        case WitnessTransaction(version, inputs, _, lockTime, witness) =>
-          WitnessTransaction(version, inputs, newOutputs, lockTime, witness)
-      }
+    val passChecksT = passInOutChecksT.flatMap { _ =>
+      TxUtil.sanityChecks(isSigned = false, inputInfos, expectedFeeRate, tx)
     }
 
-    txF.flatMap { tx =>
-      val passInOutChecksT =
-        NonInteractiveWithChangeFinalizer.sanityDestinationChecks(
-          expectedOutPoints = inputs.map(_.previousOutput),
-          expectedOutputs = outputs,
-          changeSPK = changeSPK,
-          finalizedTx = tx)
-
-      val passChecksT = passInOutChecksT.flatMap { _ =>
-        TxUtil.sanityChecks(isSigned = false,
-                            inputInfos = inputInfos,
-                            expectedFeeRate = feeRate,
-                            tx = tx)
-      }
-
-      Future.fromTry(passChecksT.map(_ => tx))
-    }
+    Future.fromTry(passChecksT.map(_ => tx))
   }
 }
 
-object NonInteractiveWithChangeFinalizer {
+object SanityCheckFinalizer {
 
   /** Checks that a finalized transaction contains the expected
     * inputs and outputs. */
   def sanityDestinationChecks(
       expectedOutPoints: Vector[TransactionOutPoint],
       expectedOutputs: Vector[TransactionOutput],
-      changeSPK: ScriptPubKey,
+      changeSPKs: Vector[ScriptPubKey],
       finalizedTx: Transaction): Try[Unit] = {
     //make sure we send coins to the appropriate destinations
     val isMissingDestination =
       !expectedOutputs.forall(finalizedTx.outputs.contains)
-    val hasExtraOutputs =
-      if (finalizedTx.outputs.size == expectedOutputs.size) {
-        false
-      } else {
-        //the extra output should be the changeOutput
-        !(finalizedTx.outputs.size == (expectedOutputs.size + 1) &&
-          finalizedTx.outputs.map(_.scriptPubKey).contains(changeSPK))
-      }
+    val hasExtraOutputs = !finalizedTx.outputs.forall(
+      output =>
+        expectedOutputs.contains(output) || changeSPKs.contains(
+          output.scriptPubKey))
```

Comment:
nit: could optimize this by doing:

```scala
val hasExtraOutputs = !finalizedTx.outputs.diff(expectedOutputs).forall(
      output => changeSPKs.contains(output.scriptPubKey))
```

that way you don't need to check the `expectedOutputs` twice

Reply:
Also, there is on check that every `changeSpk` is included